File: support/src/test/java/com/radioteria/support/test/io/MultiListenerOutputStreamTest.java
package com.radioteria.support.test.io;

import com.radioteria.util.io.MultiListenerOutputStream;
import org.junit.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Arrays;
import java.util.stream.IntStream;

import static org.junit.Assert.*;

public class MultiListenerOutputStreamTest {
    private byte[] getTestData() {
        return new byte[] { 0x00, 0x01, 0x02, 0x03 };
    }

    @Test
    public void writeToEmptyStream() {
        byte[] bytesToWrite = getTestData();

        MultiListenerOutputStream os = new MultiListenerOutputStream();
        os.write(bytesToWrite);

        assertEquals(bytesToWrite.length, os.getBytesWritten());
    }

    @Test
    public void writeToOneListener() {
        byte[] bytesToWrite = getTestData();

        ByteArrayOutputStream listener = new ByteArrayOutputStream();
        MultiListenerOutputStream os = new MultiListenerOutputStream();
        os.addListener(listener);

        os.write(bytesToWrite);

        assertEquals(bytesToWrite.length, os.getBytesWritten());
        assertArrayEquals(bytesToWrite, listener.toByteArray());
    }

    @Test
    public void writeToMoreListeners() {
        byte[] bytesToWrite = getTestData();

        ByteArrayOutputStream[] listeners = IntStream
                .range(0, 16)
                .mapToObj(n -> new ByteArrayOutputStream())
                .toArray(ByteArrayOutputStream[]::new);

        MultiListenerOutputStream os = new MultiListenerOutputStream();

        Arrays.stream(listeners).forEach(os::addListener);

        os.write(bytesToWrite);

        assertEquals(bytesToWrite.length, os.getBytesWritten());
        Arrays.stream(listeners)
                .map(ByteArrayOutputStream::toByteArray)
                .forEach(arr -> assertArrayEquals(bytesToWrite, arr));
    }

    @Test
    public void writeIfOneListenerThrowsException() {
        byte[] bytesToWrite = getTestData();

        MultiListenerOutputStream os = new MultiListenerOutputStream();

        ByteArrayOutputStream goodListener = new ByteArrayOutputStream();
        ByteArrayOutputStream badListener = new ByteArrayOutputStream() {
            @Override
            public void write(byte[] b) throws IOException {
                throw new IOException("I am a bad boy!");
            }
        };

        os.addListener(goodListener);
        os.addListener(badListener);

        assertEquals(2L, os.getListenersCount());

        os.write(bytesToWrite);

        assertEquals(1L, os.getListenersCount());
        assertArrayEquals(bytesToWrite, goodListener.toByteArray());
    }
}

File: src/test/kotlin/com/radioteria/test/util/MultiListenerOutputStreamTest.kt
package com.radioteria.test.util

import com.radioteria.util.io.MultiListenerOutputStream
import org.junit.Test

import java.io.ByteArrayOutputStream
import java.io.IOException

import org.junit.Assert.*

class MultiListenerOutputStreamTest {
    private val testData: ByteArray = byteArrayOf(0x00, 0x01, 0x02, 0x03)

    @Test
    fun writeToEmptyStream() {
        val bytesToWrite = testData

        val os = MultiListenerOutputStream()
        os.write(bytesToWrite)

        assertEquals(bytesToWrite.size.toLong(), os.bytesWritten)
    }

    @Test
    fun writeToOneListener() {
        val bytesToWrite = testData

        val listener = ByteArrayOutputStream()
        val os = MultiListenerOutputStream()
        os.addListener(listener)

        os.write(bytesToWrite)

        assertEquals(bytesToWrite.size.toLong(), os.bytesWritten)
        assertArrayEquals(bytesToWrite, listener.toByteArray())
    }

    @Test
    fun writeToMoreListeners() {
        val bytesToWrite = testData

        val listeners = (0..16)
                .map { ByteArrayOutputStream() }
                .toTypedArray()

        val os = MultiListenerOutputStream()

        listeners.forEach { os.addListener(it) }

        os.write(bytesToWrite)

        assertEquals(bytesToWrite.size.toLong(), os.bytesWritten)

        listeners
                .map { it.toByteArray() }
                .forEach { arr -> assertArrayEquals(bytesToWrite, arr) }
    }

    @Test
    fun writeIfOneListenerThrowsException() {
        val bytesToWrite = testData

        val os = MultiListenerOutputStream()

        val goodListener = ByteArrayOutputStream()
        val badListener = object : ByteArrayOutputStream() {
            @Throws(IOException::class)
            override fun write(b: ByteArray) {
                throw IOException("I am a bad boy!")
            }
        }

        os.addListener(goodListener)
        os.addListener(badListener)

        assertEquals(2L, os.listenersCount)

        os.write(bytesToWrite)

        assertEquals(1L, os.listenersCount)
        assertArrayEquals(bytesToWrite, goodListener.toByteArray())
    }
}
